### PR TITLE
feat(storage-box): implement subaccount resource and data sources

### PR DESCRIPTION
**Pull request by apricote**
Add the new resources `hcloud_storage_box_subaccount`, `data.hcloud_storage_box_subaccount` and
`data.hcloud_storage_box_subaccounts` with the CRUD endpoints:

- https://docs.hetzner.cloud/reference/hetzner#storage-box-subaccounts-list-subaccounts
- https://docs.hetzner.cloud/reference/hetzner#storage-box-subaccounts-create-a-subaccount
- https://docs.hetzner.cloud/reference/hetzner#storage-box-subaccounts-get-a-subaccount
- https://docs.hetzner.cloud/reference/hetzner#storage-box-subaccounts-update-a-subaccount
- https://docs.hetzner.cloud/reference/hetzner#storage-box-subaccounts-delete-a-subaccount

Related to #1144.

### CODE REVIEW

File: examples/data-sources/hcloud_storage_box_subaccount/data-source.tf [hunk 1]
```diff
@@ -0,0 +1,16 @@
+variable "storage_box_id" {}
+
+data "hcloud_storage_box_subaccount" "by_id" {
+  storage_box = var.storage_box_id
+  id          = 2
+}
+
+data "hcloud_storage_box_subaccount" "by_username" {
+  storage_box = var.storage_box_id
+  username    = "2025-02-12T11-35-19"
```
Comment (jooola) [inline examples/data-sources/hcloud_storage_box_subaccount/data-source.tf]: This example value is probably confusing.

File: internal/storageboxsubaccount/data_source.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+					MarkdownDescription: "Whether the Subaccount is read-only.",
+					Computed:            true,
+				},
+			},
+		},
+		"labels": schema.MapAttribute{
+			MarkdownDescription: "User-defined [labels](https://docs.hetzner.cloud/reference/cloud#labels) (key-value pairs) for the resource.",
+			ElementType:         types.StringType,
+			Computed:            true,
+		},
```
Comment (jooola) [inline internal/storageboxsubaccount/data_source.go]: Didn't we have a helper to define this?

File: internal/storageboxsubaccount/data_source_list_test.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+		Raw: fmt.Sprintf(`
+			access_settings = {
+				reachable_externally = false
+				samba_enabled = true
+				ssh_enabled = false
+				webdav_enabled = true
+				readonly = false
+			}
+
+			depends_on = [%s]`, res1.TFID()), // Only one subaccount may be created at the same time
```
Comment (jooola) [inline internal/storageboxsubaccount/data_source_list_test.go:72]: For snapshot I would understand, but for subaccounts, I think we should handle the error in the provider.

Or is this a left over from the copy pasta from snapshots?

File: internal/storageboxsubaccount/resource.go [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+	result, _, err := r.client.StorageBox.CreateSubaccount(ctx, storageBox, opts)
+	if err != nil {
+		resp.Diagnostics.Append(hcloudutil.APIErrorDiagnostics(err)...)
+		return
+	}
+
+	// Make sure to save the ID immediately so we can recover if the process stops after
+	// this call. Terraform marks the resource as "tainted", so it can be deleted and no
+	// surprise "duplicate resource" errors happen.
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root("id"), types.Int64Value(result.Subaccount.ID))...)
```
Comment (jooola) [inline internal/storageboxsubaccount/resource.go:240]: Should we save the parent storage box ID too? e.g. the user removes the resource from their config?

**Review by jooola — APPROVED** [review]

File: internal/storageboxsubaccount/data_source_list_test.go (hunk 3 above)
Comment (apricote) [reply internal/storageboxsubaccount/data_source_list_test.go:72]: Nope, I tested it without the `depends_on`, but it broke in the same way. If its fine for you, I will add it in a follow up PR for both resources.

File: internal/storageboxsubaccount/data_source.go (hunk 2 above)
Comment (apricote) [reply internal/storageboxsubaccount/data_source.go]: Only for resources :( Not for the `datasource/schema` types.

File: examples/data-sources/hcloud_storage_box_subaccount/data-source.tf (hunk 1 above)
Comment (apricote) [reply examples/data-sources/hcloud_storage_box_subaccount/data-source.tf]: Agreed, seems I copied that from the snapshot examples (where it matches the actual snapshot names).

File: internal/storageboxsubaccount/data_source_list_test.go (hunk 3 above)
Comment (jooola) [reply internal/storageboxsubaccount/data_source_list_test.go:72]: Sure

File: internal/storageboxsubaccount/data_source.go (hunk 2 above)
Comment (apricote) [reply internal/storageboxsubaccount/data_source.go]: https://github.com/hetznercloud/terraform-provider-hcloud/pull/1281

File: internal/storageboxsubaccount/resource.go (hunk 4 above)
Comment (apricote) [reply internal/storageboxsubaccount/resource.go:240]: Good catch, I assumed terraform would automagically add it to the state, as its a required attribute, but in a test the state only contained was `storage_box_id = null`, which caused the resource to be deleted from state as it could not be read, so it leaked.

Will add the attribute here and check the other resources for the same bug.

**Comment by apricote** [comment]
Failed e2e tests are because of missing server capacity, the storage box tests were successfull.
